bindings/java/src/main/com/apple/foundationdb/RangeQuery.java

```diff
@@ -213,11 +219,19 @@ private synchronized void startNextFetch() {
 			fetchOutstanding = true;
 			nextChunk = null;
 
-			fetchingChunk = tr.getRange_internal(begin, end,
-					rowsLimited ? rowsRemaining : 0, 0, streamingMode.code(),
+			nextFuture = new CompletableFuture<>();
+			if (eventKeeper != null) {
+				eventKeeper.increment(Events.RANGE_QUERY_FETCHES);
+				final long sTime = System.nanoTime();
+				nextFuture = nextFuture.thenApply((bool) -> {
```

Comment:
just a question: do we want to explicitly specify that thenApply will be called on the executor that is provided (if provided) to the Database, or would it already happen implicitly on this executor, since the nextFuture here is resolved by the code running on that executor?

I am just a bit paranoid about where future continuation runs. 
We had various cases in Snowflake where a non-careful code would end up running on ForkJoin, and that is a big no-no.

Reply:
hmmm..I'm not sure I have a huge preference here. I'd be willing to accept an alternative approach, if you have suggestions

Reply:
I meant to pass an executor to thanAsycn like here: https://github.com/apple/foundationdb/blob/master/bindings/java/src/main/com/apple/foundationdb/NativeFuture.java#L52

Not sure how easy it is to pump it from where we get it (Database constructor) to here.

Reply:
ah, I see.

For just tracking a single value, it seems wasteful to risk extra threads at all. I devised an alternative approach where we compose the `thenComplete` call so that the `FetchComplete` logic also handles the timing. The exact wall time will be slightly different, but that avoids any worry about thread management by latching to a thread that was already being used for real work. I think that should resolve this concern effectively.

Reply:
I guess this was changed, but I don't think we want to specify an executor here. The callbacks like `.thenApply` will always execute on either the thread where the future completes or on the thread that created the callback (if the future was already completed when the callback was finished), so it would only ever put work on the fork join pool if the work somehow wound up on the fork join pool by some other means.

The callbacks that all end in `Async` (e.g., `thenApplyAsync`) will schedule the callbacks for other threads. If you specify an executor, it hands off the execution to that executor; if you don't specify one, then it will schedule it on the fork join pool. So I think the rule of thumb should be something like "if you don't want the work scheduled on the completing thread, use the specified executor (to avoid scheduling on the fork join pool); if you are okay with it running not the completing thread, just use `.thenApply`". In this case, we're talking about some fairly quick `EventKeeper` updates, so scheduling on the same thread seems fine.